### ISSUE TITLE
RHS arrow function termination issues

### ISSUE THREAD

**Issue by lightmare** [valtech-nyc/proposal-fsharp-pipelines#12]
#### 1. ad Impact on precedence
> Once inside a pipeline, arrow functions are terminated at the first operator, ...

I suppose you meant "at the first pipeline operator". Because otherwise anything but primary expr would need extra parentheses:
```js
const updateScore = score =>
  score
    |> double
    |> n => n + 7
    |> n => n < 0 ? 0 : n > 100 ? 100 : n

// would parse as this nonsense
const updateScore = score =>
  (
  score
    |> double
    |> ((n => n) + 7)
    |> ((n => n) < 0)
  ) ? 0 : n > 100 ? 100 : n
```

#### 2. unexpected parsing errors
I don't know whether that's an issue with the proposal, or the implementation in Babel, but it seems to break even inside block, which it definitely shouldn't:
```js
// this currently fails to parse:
const gCached = x => x |> y => {
  return cache[y] ??= y |> much |> work
}

// this as well:
const hCached = x => x |> y => { 
  return cache[y] == null ? y |> much |> work : cache[y] 
}
```

#### 3. the whole rule feels wrong
Comma is kinda special, but imagine any other operator dictating where an arrow function ends:
```js
// what if this:
f || x => x || 5 

// parsed as:
f || (x => x) || 5
```

Here's how this early arrow function termination can be quite surprising, if you copy an existing arrow function and append it to another pipeline:
```js
let much = x => x + 1;
let work = x => x * 2;
let cache = {};

const fCached =           y => cache[y] ??= y |> much |> work;
const gCached = x => x |> y => cache[y] ??= y |> much |> work;

// populate cache
expect(fCached(3)).toBe(8);
expect(cache[3]).toBe(8);
expect(gCached(5)).toBe(12);
expect(cache[5]).toBe(12); // fails, it cached 5, not the result

work = x => "should've used cache";
expect(fCached(3)).toBe(8); // returns cached result, doesn't call work
expect(gCached(3)).toBe(8); // fails because it calls work
```

**Comment by mAAdhaTTah** [comment]
> I suppose you meant "at the first pipeline operator".

This is correct. I'm not sure I follow the issues you're raising though. My expectation is the things you're suggesting fail behave the way you expect and I think it's more likely there's a bug in Babel, possibly related to the implementation of the new assignment operators (or assignment in general).

**Comment by lightmare** [comment]
I've reported the parsing issues in Babel repo.

Let me clarify the 3rd point, which to me seems like an issue with the design, not the implementation. Due to the "terminates at the first pipeline" rule, the pipeline operator can terminate an expression of lower precedence.

```js
x |> y => cache[y] = y |> work
// parses as:
x |> (y => (cache[y] = y)) |> work
```

Is it acceptable that `|>` can sometimes terminate an assignment, despite `|>` binding more tightly than assignment in other contexts?
